File: ts/longestCommonPrefix.ts
function longestCommonPrefix(strs: string[]): string {
  const first = strs[0];
  let cache = '';

  for (let i = 0; i < first.length; i++) {
    const prefix = first.slice(0, i + 1);
    const isSame = strs.every(it => {
      const compare = it.slice(0, i + 1);
      return prefix === compare;
    });

    console.log({ prefix, isSame });

    if (isSame) cache = prefix;
    else return cache;
  }

  return cache;
}

export default () => {
  // const strs = ['flower', 'flow', 'flight'];
  // const strs = ['dog', 'racecar', 'car'];
  const strs = ['a'];
  const result = longestCommonPrefix(strs);

  console.log(result);
};

File: rs/src/longest_common_prefix.rs
fn longest_common_prefix(strs: Vec<String>) -> String {
    let mut sorted = strs.to_vec();
    sorted.sort_by(|a, b| a.len().cmp(&b.len()));

    let first = &sorted[0];
    println!("{:?}", sorted);

    let mut cache = "";

    for i in 0..first.len() {
        let prefix = &first[0..i + 1];
        let is_same = sorted.iter().all(|x| &x[0..i + 1] == prefix);

        if is_same {
            cache = prefix;
        } else {
            return cache.to_string();
        }
    }

    return cache.to_string();
}

pub fn main() {
    let strs: Vec<String> = vec![String::from("ab"), String::from("a")];
    let result = longest_common_prefix(strs);
    println!("{result}");
}
